Panic in S3 Select

<!--- Provide a general summary of the issue in the Title above -->

## Expected Behavior
<!--- If you're describing a bug, tell us what should happen -->
<!--- If you're suggesting a change/improvement, tell us how it should work -->

MinIO should not crash

## Current Behavior
<!--- If describing a bug, tell us what happens instead of the expected behavior -->
<!--- If suggesting a change/improvement, explain the difference from current behavior -->

MinIO crashes when running the same query again with mc when we interrupt it the first time.

## Steps to Reproduce (for bugs)
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->
1. `mc sql -e 'select * from s3object s ' --json-input "Type=LINES"  minio/bucket/some.json`
2. Hit Ctrl+C on mc before mc finishes its output.
3. Run same command again.
4. minIO server crashes with:

```
panic: read /home/aditya/data/bucket/some.json: file already closed

goroutine 211 [running]:
github.com/bcicen/jstream.newScanner.func1(0xc000054a00, 0x7f4d03be2ec0, 0xc000485ca0)
        /q/.q/sources/gopath/pkg/mod/github.com/bcicen/jstream@v0.0.0-20190220045926-16c1f8af81c2/scanner.go:48 +0x180
created by github.com/bcicen/jstream.newScanner
        /q/.q/sources/gopath/pkg/mod/github.com/bcicen/jstream@v0.0.0-20190220045926-16c1f8af81c2/scanner.go:32 +0xe5
```

## Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

Testing s3 select.

## Regression
<!-- Is this issue a regression? (Yes / No) -->
<!-- If Yes, optionally please include minio version or commit id or PR# that caused this regression, if you have these details. -->

Most likely yes, but not sure.

## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->

```
Version: 2019-10-12T01:39:57Z
Release-Tag: RELEASE.2019-10-12T01-39-57Z
Commit-ID: bd106408462ecef70debf51f1e6179de950c5812
```

It seems like a generic race when closing the streams. On one of my runs I got this:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x8 pc=0x110f4cc]

goroutine 19939 [running]:
github.com/klauspost/readahead.(*buffer).buffer(...)
        e:/gopath/pkg/mod/github.com/klauspost/readahead@v1.3.0/reader.go:385
github.com/klauspost/readahead.(*reader).Read(0xc000397180, 0xc0006f3020, 0xfff, 0xfff, 0xc0003f5800, 0xc000e4bf10, 0x4073ff)
        e:/gopath/pkg/mod/github.com/klauspost/readahead@v1.3.0/reader.go:256 +0xac
github.com/minio/minio/cmd.(*GetObjectReader).Read(0xc0006f0000, 0xc0006f3020, 0xfff, 0xfff, 0xc0005d2088, 0xc0005d2138, 0xc0000447d8)
        e:/gopath/src/github.com/minio/minio/cmd/object-api-utils.go:709 +0x63
github.com/minio/minio/pkg/s3select.(*countUpReader).Read(0xc000175f00, 0xc0006f3020, 0xfff, 0xfff, 0xc000083020, 0xc000e4bf88, 0x1120101)
        e:/gopath/src/github.com/minio/minio/pkg/s3select/progress.go:34 +0x58
github.com/minio/minio/pkg/s3select.(*countUpReader).Read(...)
        e:/gopath/src/github.com/minio/minio/pkg/s3select/progress.go:34
github.com/minio/minio/pkg/s3select.(*progressReader).Read(0xc000175f60, 0xc0006f3020, 0xfff, 0xfff, 0x100, 0x0, 0x0)
        e:/gopath/src/github.com/minio/minio/pkg/s3select/progress.go:56 +0x65
github.com/bcicen/jstream.newScanner.func1(0xc0006f2000, 0x9be2658, 0xc000175f60)
        e:/gopath/pkg/mod/github.com/bcicen/jstream@v0.0.0-20190220045926-16c1f8af81c2/scanner.go:37 +0x7f
created by github.com/bcicen/jstream.newScanner
        e:/gopath/pkg/mod/github.com/bcicen/jstream@v0.0.0-20190220045926-16c1f8af81c2/scanner.go:32 +0xec
```

Seems related and looks like a use after close. Edit: Confirmed, this is a read after Close.

Seems like jstream is particularly nasty: https://github.com/bcicen/jstream/blob/master/scanner.go#L48

--- a/go.mod
+++ b/go.mod
@@ -32,7 +32,7 @@ require (
 	github.com/json-iterator/go v1.1.7
 	github.com/klauspost/compress v1.8.3
 	github.com/klauspost/pgzip v1.2.1
-	github.com/klauspost/readahead v1.3.0
+	github.com/klauspost/readahead v1.3.1
 	github.com/klauspost/reedsolomon v1.9.3
 	github.com/kurin/blazer v0.5.4-0.20190613185654-cf2f27cc0be3
 	github.com/lib/pq v1.0.0

--- a/pkg/s3select/json/reader.go
+++ b/pkg/s3select/json/reader.go
@@ -17,7 +17,9 @@
 package json
 
 import (
+	"errors"
 	"io"
+	"sync"
 
 	"github.com/minio/minio/pkg/s3select/sql"
 
@@ -67,7 +69,6 @@ func (r *Reader) Read(dst sql.Record) (sql.Record, error) {
 // Close - closes underlying reader.
 func (r *Reader) Close() error {
 	// Close the input.
-	// Potentially racy if the stream decoder is still reading.
 	err := r.readCloser.Close()
 	for range r.valueCh {
 		// Drain values so we don't leak a goroutine.
@@ -78,6 +79,7 @@ func (r *Reader) Close() error {
 
 // NewReader - creates new JSON reader using readCloser.
 func NewReader(readCloser io.ReadCloser, args *ReaderArgs) *Reader {
+	readCloser = &syncReadCloser{rc: readCloser}
 	d := jstream.NewDecoder(readCloser, 0).ObjectAsKVS()
 	return &Reader{
 		args:       args,
@@ -85,4 +87,47 @@ func NewReader(readCloser io.ReadCloser, args *ReaderArgs) *Reader {
 		valueCh:    d.Stream(),
 		readCloser: readCloser,
 	}
+}
+
+// syncReadCloser will wrap a readcloser and make it safe to call Close
+// while reads are running.
+// All read errors are also postponed until Close is called and
+// io.EOF is returned instead.
+type syncReadCloser struct {
+	rc    io.ReadCloser
+	errMu sync.Mutex
+	err   error
+}
+
+func (pr *syncReadCloser) Read(p []byte) (n int, err error) {
+	// This ensures that Close will block until Read has completed.
+	// This allows another goroutine to close the reader.
+	pr.errMu.Lock()
+	defer pr.errMu.Unlock()
+	if pr.err != nil {
+		return 0, io.EOF
+	}
+	n, pr.err = pr.rc.Read(p)
+	if pr.err != nil {
+		// Translate any error into io.EOF, so we don't crash:
+		// https://github.com/bcicen/jstream/blob/master/scanner.go#L48
+		return n, io.EOF
+	}
+
+	return n, nil
+}
+
+var errClosed = errors.New("read after close")
+
+func (pr *syncReadCloser) Close() error {
+	pr.errMu.Lock()
+	defer pr.errMu.Unlock()
+	if pr.err == errClosed {
+		return nil
+	}
+	if pr.err != nil {
+		return pr.err
+	}
+	pr.err = errClosed
+	return pr.rc.Close()
 }

--- a/pkg/s3select/json/reader_test.go
+++ b/pkg/s3select/json/reader_test.go
@@ -33,22 +33,43 @@ func TestNewReader(t *testing.T) {
 		t.Fatal(err)
 	}
 	for _, file := range files {
-		f, err := os.Open(filepath.Join("testdata", file.Name()))
-		if err != nil {
-			t.Fatal(err)
-		}
-		r := NewReader(f, &ReaderArgs{})
-		var record sql.Record
-		for {
-			record, err = r.Read(record)
+		t.Run(file.Name(), func(t *testing.T) {
+			f, err := os.Open(filepath.Join("testdata", file.Name()))
 			if err != nil {
-				break
+				t.Fatal(err)
 			}
-		}
-		r.Close()
-		if err != io.EOF {
-			t.Fatalf("Reading failed with %s, %s", err, file.Name())
-		}
+			r := NewReader(f, &ReaderArgs{})
+			var record sql.Record
+			for {
+				record, err = r.Read(record)
+				if err != nil {
+					break
+				}
+			}
+			r.Close()
+			if err != io.EOF {
+				t.Fatalf("Reading failed with %s, %s", err, file.Name())
+			}
+		})
+
+		t.Run(file.Name()+"-close", func(t *testing.T) {
+			f, err := os.Open(filepath.Join("testdata", file.Name()))
+			if err != nil {
+				t.Fatal(err)
+			}
+			r := NewReader(f, &ReaderArgs{})
+			r.Close()
+			var record sql.Record
+			for {
+				record, err = r.Read(record)
+				if err != nil {
+					break
+				}
+			}
+			if err != io.EOF {
+				t.Fatalf("Reading failed with %s, %s", err, file.Name())
+			}
+		})
 	}
 }
 

--- a/pkg/s3select/progress.go
+++ b/pkg/s3select/progress.go
@@ -18,8 +18,10 @@ package s3select
 
 import (
 	"compress/bzip2"
+	"errors"
 	"fmt"
 	"io"
+	"sync"
 	"sync/atomic"
 
 	gzip "github.com/klauspost/pgzip"
@@ -50,13 +52,29 @@ type progressReader struct {
 	rc              io.ReadCloser
 	scannedReader   *countUpReader
 	processedReader *countUpReader
+
+	closedMu sync.Mutex
+	closed   bool
 }
 
 func (pr *progressReader) Read(p []byte) (n int, err error) {
+	// This ensures that Close will block until Read has completed.
+	// This allows another goroutine to close the reader.
+	pr.closedMu.Lock()
+	defer pr.closedMu.Unlock()
+	if pr.closed {
+		return 0, errors.New("progressReader: read after Close")
+	}
 	return pr.processedReader.Read(p)
 }
 
 func (pr *progressReader) Close() error {
+	pr.closedMu.Lock()
+	defer pr.closedMu.Unlock()
+	if pr.closed {
+		return nil
+	}
+	pr.closed = true
 	return pr.rc.Close()
 }
 

